### ISSUE TITLE
Techempower Web Framework Benchmarks: colossus added

### ISSUE THREAD

**Issue by mfirry** [tumblr/colossus#210]
Hi, 
don't know if you guys heard of [Techempower's Web Framework Benchmarks](https://www.techempower.com/benchmarks). 
They basically have been running some tests and compared different web frameworks. 

Their 11th round will be up soon and I've added [code for colossus](https://github.com/TechEmpower/FrameworkBenchmarks/tree/master/frameworks/Scala/colossus). Preliminary results are already available [here](https://www.techempower.com/benchmarks/previews/round11/). 

Their second preview of round11 will be available on June 16th. If you guys feel like my code is missing something, feel free to PR.

**Comment by nsauro** [comment]
Any reason why Colossus is only showing up in the JSON section, and not any others?  

If we wanted to swap out lift-json for something else(ie: jackson), would that be cool?

**Comment by nsauro** [comment]
Also, I'm curious to the 74 errors in the plaintext section.  Anyway we can have a peek at those?

**Comment by mfirry** [comment]
Just to make things clear: I am in no way involved in Techempower or anything. I was just passively following their benchmark until I realized I wanted to see results for frameworks I follow/like (colossus*, http4s). So I did what they [require](https://www.techempower.com/benchmarks/#section=code) you to do if you want to add your own implementation.

That said, you can either fork their repo and fix/add what I already committed.
Or we can work together, should you like us to do.

errors for the preview tests are [here](https://github.com/TechEmpower/TFB-Round-11/tree/master/peak/linux/latest/logs/colossus)

There's a google group if you want to talk to them [here](https://groups.google.com/forum/?fromgroups=#!forum/framework-benchmarks).

*I do realize colossus is much more than "just" a web framework, but nonetheless I wanted to check out on its performances.

**Comment by nsauro** [comment]
Ah..ok, my mistake.  Well, thanks for the heads up, this is actually something we've been meaning to do, so thanks for spurring us on!

**Comment by mfirry** [comment]
Sure, no problem. Let me know if you go ahead with this and if you need me for anything. (you can close this non-issue)

**Comment by DanSimon** [comment]
Hey thanks a bunch for submitting the code.  I was actually going to do this myself pretty soon.  The service code itself looks pretty good.

I'll have to dive in a bit and see what's going on.  The results from both benchmarks diverge quite a bit from our internal benchmarks (done on similar hardware and in the same fashion), but the throughput on the plaintext benchmark is especially surprising.  In all our internal benchmarks Colossus is basically on par with Vertx, so I would expect to see similar results here.

Also the fact that the plaintext latency numbers for Colossus are significantly _lower_ than everybody else is really bizarre.  It leads me to believe that the service was not actually being pushed to its limits.  Also all of the "dropped reply" error messages are very odd.

**Comment by DanSimon** [comment]
Ok, so I've uncovered a few very interesting things.  For one thing, I've been using wrk wrong this whole time!  I had assumed it was doing pipelining on its own, and didn't realize you had to feed it a Lua script to enable it (which in retrospect makes a whole lot more sense actually).  you can see where Techempower is doing it [here](https://github.com/TechEmpower/FrameworkBenchmarks/blob/master/toolset/benchmark/framework_test.py#L108)

In short, I found a pretty big performance bottleneck in Colossus with pipelined requests.  Essentially it's being way too eager with trying to write responses to the socket.  Currently it's doing a write for every response, whereas it should be buffering and doing only one write per event loop iteration.  I experimented with a hacky implementation that adds the buffering and that immediately more than doubled performance.

I have a similar setup as the Techempower environment (2 machines with dual Xeon E5 v2 and 10GbE) and now I'm getting 2.1M req/sec (with latency 50% at 1.2ms) for a barebones Colossus service.  

I'll be working on the proper implementation for the fix over the next couple days and we'll push it out with an 0.6.4 release and hopefully get it in before the next preliminary run on the 16th.

In terms of the benchmark code, I also found a few things and will be submitting a PR once we get the updates into Colossus:
* generating a timestamp per request has a huge performance impact.  We can use an actor with a timer to generate them only once per second.  Eventually we should probably bake that logic into colossus.
* chaining `.withHeader` calls also has a performance hit.  We should just create the `HttpResponse` all in one shot
* The Url matching DSL is doing some somewhat inefficient stuff.  I've known about this but the impact is more than I would have suspected.  I think for the benchmark we should just look at the raw url.

There's still some weirdness with the Techempower numbers, particularly the latency numbers, but I think this'll put us on the right track.  So I'll leave this ticket open until we get all that done and merged in.  I'll also start a separate issue for the buffering fix.

**Comment by mfirry** [comment]
Awesome work! I don't know whether I should be happy or sorry this all thing came out somehow because of me.

Keep it goin' folks!

**Comment by mfirry** [comment]
Preliminary results: https://www.techempower.com/benchmarks/previews/round11/

**Comment by DanSimon** [comment]
Gonna close this issue now as I think all the low-hanging fruit has been addressed.  The first round of fixes nearly doubled our score on the plaintext benchmark and another PR is out that I think will give it another big bump.  On our own hardware, I have the benchmark service up from 2.1M req/sec to 3.6M.
